refactor: replace deprecated String.prototype.substr()

CommanderRoot opened the pull request:
[String.prototype.substr()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr) is deprecated so we replace it with [String.prototype.slice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice) or [String.prototype.substring()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring) which work similarily but aren't deprecated.
.substr() probably isn't going away anytime soon but the change is trivial so it doesn't hurt to do it.

poltak commented:
This looks good! Thanks for the contribution.
One thing I noticed was your two replacements of `.substr` with `.substring` still contain the exact same args, even though the two functions treat the second arg differently. Though, from a quick read of the surrounding code, it seems like the original author might have thought `.substr` works the same as `.substring` (or just confused them), hence that second arg doesn't need to be changed? Is that how you interpreted it?

CommanderRoot commented:
The reason I used `substring()` [instead](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring#the_difference_between_substring_and_substr) of `slice()` in these instances is because the second parameter is normally positive but it **could** be negative.
Example:
`
"0123456789".substr(0, -1) == ""
"0123456789".substring(0, -1) == ""
"0123456789".slice(0, -1) == "012345678"
`
So to replicate the current behaviour I used `substring()`.

iamtouchskyer commented:
## Code Review: Replace deprecated String.prototype.substr()

**Summary**: Modernizing codebase by replacing deprecated substr() calls with slice() and substring().

### ✅ Changes Analysis

All replacements are **functionally correct** and improve code quality:

1. **substr(0, prefix.length) → slice(0, prefix.length)**: ✅ Identical behavior
2. **substr(0, current.selection[cursorToMove] - 1) → substring(0, current.selection[cursorToMove] - 1)**: ✅ Good choice for potentially negative indices
3. **substr(current.selection[cursorToMove] + 1) → slice(current.selection[cursorToMove] + 1)**: ✅ Safe slice from middle
4. **Multiple similar patterns**: ✅ All consistently applied
5. **PDF anchoring cleanup**: ✅ Improved by removing unnecessary length variable and using slice(start, end) directly

### 🎯 Benefits

- **Future-proofing**: Removes deprecated API usage
- **Consistency**: Uses modern string methods throughout
- **No behavioral changes**: All replacements maintain exact functionality
- **Clean code**: The PDF.js change actually improves readability

### 🔍 Testing Recommendation

While these are low-risk changes, testing should focus on:
- Text selection functionality (most changes are in SelectionModifiers)
- PDF highlighting/anchoring
- Notification text truncation
- Backup restore procedures

### ✅ Recommendation: **APPROVE** 

This is a solid refactoring PR that:
- Removes deprecated API usage
- Maintains backward compatibility  
- Improves code maintainability
- Has been waiting too long for review (2+ years!)

Great contribution @CommanderRoot! 🙏

iamtouchskyer commented:
🔧 **CODE QUALITY IMPROVEMENT - Needs Rebase**

**Summary**: Replaces deprecated `String.prototype.substr()` with modern alternatives

✅ **Good Changes**:
- Replaces deprecated `substr()` with `slice()` and `substring()`
- Future-proofs codebase against potential removal of deprecated method
- Maintains identical functionality with safer alternatives

❌ **Blocking Issue**: CONFLICTING status - merge conflicts need resolution

**Technical Review**:
The changes look good and follow best practices:
- `substr()` → `slice()` for simple substring extraction
- `substr()` → `substring()` where appropriate
- No functional changes, only API modernization

**Required Action**: Please rebase against latest main branch to resolve conflicts:
```bash
git rebase origin/main
# resolve conflicts
git push --force-with-lease
```

**Recommendation**: APPROVE after conflicts resolved ✅

CommanderRoot committed:
refactor: replace deprecated String.prototype.substr()

.substr() is deprecated so we replace it with .slice() or .substring() which work similarily but aren't deprecated

Signed-off-by: Tobias Speicher <rootcommander@gmail.com>

iamtouchskyer reviewed (approved):
✅ **Excellent refactoring work!**

This is exactly the kind of proactive code maintenance that keeps the codebase healthy. Replacing deprecated `String.prototype.substr()` with modern alternatives is a great improvement.

**What I reviewed:**
- ✅ All changes correctly replace `.substr()` with `.slice()` or `.substring()`
- ✅ Logic remains functionally identical 
- ✅ Affects 6 files with clean, focused changes
- ✅ No breaking changes or side effects
- ✅ Good commit message explaining the reasoning

**Technical notes:**
- The replacements are semantically correct
- `.slice()` is generally preferred for most use cases
- Great to see the commit is signed and follows conventional commit format

This PR is ready to merge! Thanks @CommanderRoot for taking initiative on code quality improvements. 🚀
